### ISSUE TITLE
torch.nn.functional.normalize epsilon too small for half precision

### ISSUE THREAD

**Issue by rmrao** [pytorch/pytorch#32137]
`torch.nn.functional.normalize` uses an epsilon of 1e-12. This value is too small for half precision and gets evaluated as zero (`torch.HalfTensor([1e-12]) == 0`). This causes nans in half precision when the norm is evaluated as zero. This should be fixable by increasing the epsilon to 1e-6.

**Comment by vadimkantorov** [comment]
(One idea is to leverage: torch.finfo(x.dtype).eps or sth like that inside the impl, or provide an eps argument and do this only if it's unset by the user)

**Comment by ezyang** [comment]
Increasing the epsilon seems reasonable. Send us a PR?

**Comment by uniartisan** [comment]
I'm also having the same issue, should this be the default option for torch?
In torch2.6 eps is 1e-12, which is no problem for bf16 and fp32, but fp16 may encounter nan.

**Comment by vadimkantorov** [comment]
Also, maybe normalize can be fixed to avoid NaN output even for 0 epsilon? Then this could improve the fp16 situation even without the b/c-breaking default fp16 epsilon change. 

Some alternative solutions: maybe introduce a global per-op, per-dtype default constant registry

**Comment by albanD** [comment]
Opened an umbrella issue at https://github.com/pytorch/pytorch/issues/151110
Let's update the epsilon here following the rules from the other issue!
